examples/recording_service/leveldb-integration/cpp/README.md

```diff
@@ -0,0 +1,1077 @@
+  objects will be passed to the method.
+- **`PubDiscoveryLevelDbWriter`**: this class extends the API class
+  `rti::recording::storage::PublicationStorageWriter`. This class
+  is in charge of storing samples of the built-in `DCPSPublication`
+  topic. *Recorder* will ask the `LevelDbWriter` instance to create
+  an instance of this class during startup, by calling method 
+  `create_publication_writer()`.
+  The instance of this class will be in charge of creating the 
+  `DCPSPublication.dat` database.
+  The main method this class implements is the `store()` method.
+  Available samples of topic `DCPSPublication` and their associated
+  sample info objects will be passed to the call.
+  
+#### Implementation details of class `LevelDbStreamWriter`
+
+This Storage Writer plug-in works by storing data in _serialized_
+CDR format, by using the types described in the above section
+([Format of the stored data](#format-of-the-stored-data)). The code
+for the types defined in file `LevelDb_RecorderTypes.idl` will be
+generated automatically. The generated code includes methods to
+serialize and deserialize samples to and from CDR format.
+
+The main members of the `LevelDbStreamWriter` class are:
+
+- `data_db_`: this is the LevelDB database object (leveldb::DB) that 
+  will be used to store the data.
+- `key_comparator`: the custom key comparator will be passed when
+  the creation of the database happens. The key comparator checks
+  the reception timestamp of the samples and returns them in
+  monotonic ascending order.
+- `key_buffer_`: a byte (char) buffer used to store the serialized
+  key sample.
+- `value_buffer_`: a byte (char) buffer used to store the serialized
+  value sample.
+
+**Constructor**
+
+The constructor of the class has to open the database where the
+new topic will be stored. It will construct the name of the 
+database with the stream name, taken from the `StreamInfo` object
+passed as a parameter, and the domain ID, provided as a parameter
+too.
+
+Another key aspect of the constructor is to provide a sensible
+initialization of the buffers (byte vectors) that will be used
+to serialize the samples stored in the database. This will
+minimize the number of memory allocations performed by the
+vectors.
+
+For the key buffer's initialization size, we will use the key
+type's max sample serialized size:
+
+```cpp
+const StructType& key_type = rti::topic::dynamic_type<UserDataKey>::get();
+key_buffer_.resize(key_type.cdr_serialized_sample_max_size());
+```
+
+In the case of the value buffer, the serialized sample's max size
+not only depends on the size of type `UserDataValue`, but mostly
+on the size of the topic's type's sample max serialized size. Hence,
+we first get the type from the `StreamInfo` object, which is
+represented as a native `DDS_TypeCode` pointer object. With it, we
+do the following:
+
+- If the type is unbounded, we initialize the buffer to 8 KB.
```

Comment:
why not doing always this? potentially we could have a default of, perhaps, 64KB and resize as needed. Even we could mention that this minimum configuraiton could be configurable if needed.

Reply:
It's an option. But if we can reserve memory upfront during startup, why doing it afterwards? That was my thinking.

Reply:
I agree, but I was considering that we could always follow that approach for simplicity in the code. It's not a big deal.